HW1/network_commands.py

```diff
@@ -0,0 +1,253 @@
+        self.cur_iteration = 0
+
+        self.num_unique_vs_iteration = None
+
+        # Run this at the end!
+        self.initialize_groups()
+
+    def initialize_groups(self):
+        self.connected_rdd = Connected_Components.initialize_groups_static(self.sc, self.network_rdd)
+        self.cur_iteration += 1
+
+    def do_iteration(self):
+        self.connected_rdd = Connected_Components.do_iteration_static(self.sc, self.connected_rdd)
+        self.cur_iteration += 1
+
+    def get_num_unique_groups(self):
+        list_of_indices = self.connected_rdd.map(lambda x: x[1][1])
+        num_distinct_indices = list_of_indices.distinct().count()
+        return num_distinct_indices
+
+    def get_number_of_nodes_per_index(self):
+        list_of_indices = self.connected_rdd.map(lambda x: x[1][1])
+        list_of_indices_count = list_of_indices.map(lambda x: (x, 1))
+        return list_of_indices_count.reduceByKey(lambda a, b: a + b, num_partitions)
+
+    def run_until_converged(self):
+        go = True
+
+        self.num_unique_vs_iteration = []
+
+        previous_num_groups = self.get_num_unique_groups()
+        groups_before_update = previous_num_groups
+
+        self.num_unique_vs_iteration.append(groups_before_update)
+
+        while go:
+            print 'Beginning of Iteration', self.cur_iteration
+            print 'Before update:' , groups_before_update, 'groups'
+            self.do_iteration()
+            groups_before_update = self.get_num_unique_groups()
+            self.num_unique_vs_iteration.append(groups_before_update)
+
+            if groups_before_update == previous_num_groups:
+                go = False
+            else:
+                previous_num_groups = groups_before_update
+        print 'Finished at end of iteration' , self.cur_iteration - 1 , '!'
+        print 'There are ', groups_before_update, ' groups in the network.'
+
+    #### STATIC METHODS TO INTERACT WITH SPARK ####
+    @staticmethod
+    def initialize_groups_static(sc, network_rdd):
+        labeled_network = network_rdd.zipWithIndex().map(lambda x: (x[0][0], (x[0][1], x[1])))
+        return labeled_network.partitionBy(num_partitions).cache()
+
+    @staticmethod
+    def do_iteration_static(sc, connected_rdd):
+
+        def get_children_index(x):
+            children = x[1][0]
+            index = x[1][1]
+            return [(z, index) for z in children]
+
+        def get_smaller_index(a, b):
+            if a < b:
+                return a
+            return b
+
+        children_indices = connected_rdd.flatMap(get_children_index)
+        child_with_smallest_index = children_indices.reduceByKey(get_smaller_index, num_partitions) # This is partitioned
+        joined_rdd= connected_rdd.join(child_with_smallest_index) # copartitioned!
+        connected_rdd = joined_rdd.map(lambda x: (x[0], (x[1][0][0], x[1][1])), preservesPartitioning=True) # Should preserve partitioning...
+        return connected_rdd.cache()
```

Comment:
Nice object-oriented code.  But is this the best way to code a couple variations of bfs?